### ISSUE TITLE
Tensor subclass constructor via autograd function is error prone re mutation

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#127996]
### 🐛 Describe the bug

```
import torch
from torch.utils._pytree import tree_map_only

class T(torch.Tensor):
    def __new__(cls, inner, *, requires_grad=False):
        res = torch.Tensor._make_wrapper_subclass(
            cls,
            size=inner.size(),
            strides=inner.stride(),
            storage_offset=inner.storage_offset(),
            dtype=inner.dtype,
            layout=inner.layout,
            device=inner.device,
            requires_grad=requires_grad,
        )
        res.inner = inner
        return res

    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        inner_args, inner_kwargs = tree_map_only(T, lambda t: t.inner, (args, kwargs))
        inner_res = func(*inner_args, **inner_kwargs)
        return tree_map_only(torch.Tensor, T, inner_res)

    def __repr__(self):
        return f"T({self.inner})"

class MakeT(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return T(x)

    @staticmethod
    def backward(ctx, grad_out):
        return grad_out.inner

x = torch.randn(1, requires_grad=True)
x2 = x * 2
# y = x2[:]
y = MakeT.apply(x2)
x2.mul_(3)
y.backward()
print(x.grad)
```

This prints

```
tensor([2.])
```

But if you replace `y = MakeT.apply(x2)` with `y = x2[:]`, you get 6, the correct gradient. The problem is that y is a view of x2, but autograd.Function has failed to track it.

If you try to implement this without a tensor subclass, autograd.Function will report an error saying that you can't access the grad_fn of y after mutating x2 because the grad_fn is no longer valid for custom autograd subclass. But we need to look into the inner members of generated tensor subclasses to find less obvious aliasing.

It is likely that DTensor suffers from this problem.

cc @albanD @bdhirsh 

### Versions

main
